### ISSUE TITLE
How to use ValueSerializer

### ISSUE THREAD

**Issue by arturekdev** [A248/DazzleConf#25]
Hi, how i can use ValueSerializer?

**Comment by arturekdev** [comment]
fast

**Comment by A248** [comment]
You use ValueSerialiser so you can use a custom type in your configuration. Let's say you had the following:

```java
public class CustomType {

	private final String value;

	public CustomType(String value) {
		this.value = value;
	}

	public String stringValue() {
		return value;
	}
}
```


1. First implement the ValueSerialiser:
```java
public class MyCustomSerializer implements ValueSerialiser<CustomType> {
	@Override
	public Class<CustomType> getTargetClass() {
		return CustomType.class;
	}

	@Override
	public CustomType deserialise(FlexibleType flexibleType) throws BadValueException {
		// Convert from a String -> CustomType
		String value = flexibleType.getString();
		return new CustomType(value);
	}

	@Override
	public Object serialise(CustomType value, Decomposer decomposer) {
		// Convert from a CustomType -> String
		return value.stringValue();
	}
}
```
2. Then add the serializer to the ConfigurationOptions:
```java
ConfigurationOptions options = new ConfigurationOptions.Builder()
				.addSerialiser(new MyCustomSerializer())
				.build();
```
Use the ConfigurationOptions when you create your config factory (GsonConfigurationFactory, SnakeYamlConfigurationFactory, etc.)

**Comment by arturekdev** [comment]
How i can use this for object with more variables?

**Comment by A248** [comment]
I'm not exactly sure what you mean by that. If you need to use this type in your configuration, you can very easily:

```java
public interface Config {

  @DefaultString("default value")
  CustomType something();

  // You can even use it in a List, Set, or Collection
  @DefaultStrings({"value1", "value2"})
  List<CustomType> listOfThings();

}
```

**Comment by arturekdev** [comment]
I need put this to config:

`    static Map<Integer, Level> defaultLevels() {
        Map<Integer, Level> levelsMap = new HashMap<>();
        levelsMap.put(1, new Level(5, 10, 1, 10, 0));
        levelsMap.put(2, new Level(7, 12, 2, 11, 10));
        levelsMap.put(3, new Level(9, 14, 3, 12, 20));
        levelsMap.put(4, new Level(12, 16, 4, 13, 30));
        levelsMap.put(5, new Level(15, 18, 5, 14, 40));
        return levelsMap;
    }

    @ConfDefault.DefaultObject("defaultLevels")
    Map<Integer, Level> levels();`

**Comment by A248** [comment]
I see. For that, you can treat a `Level` as a list of integers and implement your serializer like this:
```java
public class LevelSerializer implements ValueSerialiser<Level> {
	@Override
	public Class<Level> getTargetClass() {
		return Level.class;
	}

	@Override
	public Level deserialise(FlexibleType flexibleType) throws BadValueException {
		List<Integer> values = new ArrayList<>();
		for (FlexibleType element : flexibleType.getList()) {
			values.add(element.getInteger());
		}
		return new Level(values);
	}

	@Override
	public Object serialise(Level value, Decomposer decomposer) {
		return value.values();
	}
}

public class Level {

	private final List<Integer> values;

	public Level(List<Integer> values) {
		this.values = List.copyOf(values);
		if (this.values.size() != 5) {
			throw new IllegalArgumentException("Size of Level values must be 5");
		}
	}

	public List<Integer> values() {
		return values;
	}
}
```
There are other ways you could do this, like by treating a Level as a configuration section, but I choose this way because it's relatively simple.

**Comment by arturekdev** [comment]
But you can show my how i can do ths like by treating a level as configuration section? Your solution is not convenient to configuration.

**Comment by A248** [comment]
Yes. You can even do this without using a value serializer:

```java
public interface Config {

   static Map<Integer, Level> defaultLevels() {
        Map<Integer, Level> levelsMap = new HashMap<>();
        levelsMap.put(1, Level.of(5, 10, 1, 10, 0));
        levelsMap.put(2, Level.of(7, 12, 2, 11, 10));
        levelsMap.put(3, Level.of(9, 14, 3, 12, 20));
        levelsMap.put(4, Level.of(12, 16, 4, 13, 30));
        levelsMap.put(5, Level.of(15, 18, 5, 14, 40));
        return levelsMap;
    }

    @ConfDefault.DefaultObject("defaultLevels")
    Map<Integer, Level> levels();


}
public interface Level {

	int minSpawnDelay();

	int maxSpawnDelay();

	int spawnCount();

	int spawnRange();

	int price();

	static Level of(int minSpawnDelay, int maxSpawnDelay, int spawnCount, int spawnRange, int price) {
		return new Level() {

			@Override
			public int minSpawnDelay() { return minSpawnDelay; }
			@Override
			public int maxSpawnDelay() { return maxSpawnDelay; }
			@Override
			public int spawnCount() { return spawnCount; }
			@Override
			public int spawnRange() { return spawnRange; }
			@Override
			public int price() { return price; }
		};
	}

}
```

**Comment by A248** [comment]
Note that you need `@SubSection` here:
```java
    @ConfDefault.DefaultObject("defaultLevels")
    Map<Integer, @SubSection Level> levels();
```

**Comment by arturekdev** [comment]
Thank you wery much <3

**Comment by A248** [comment]
You're welcome. A thanks to you too - I've added this information to the wiki now:
https://github.com/A248/DazzleConf/wiki/Examples-of-custom-types-in-your-configuration

**Comment by arturekdev** [comment]
Map key need be String :(
```
java.lang.ClassCastException: java.lang.Integer cannot be cast to java.lang.String
        at space.arim.dazzleconf.ext.snakeyaml.CommentedWriter.writeRemainingMapEntries(CommentedWriter.java:99) ~[?:?]
        at space.arim.dazzleconf.ext.snakeyaml.CommentedWriter.writeMap(CommentedWriter.java:93) ~[?:?]
        at space.arim.dazzleconf.ext.snakeyaml.CommentedWriter.lambda$writeValue$0(CommentedWriter.java:188) ~[?:?]
        at space.arim.dazzleconf.ext.snakeyaml.CommentedWriter.descendAndDo(CommentedWriter.java:157) ~[?:?]
        at space.arim.dazzleconf.ext.snakeyaml.CommentedWriter.writeValue(CommentedWriter.java:188) ~[?:?]
        at space.arim.dazzleconf.ext.snakeyaml.CommentedWriter.writeRemainingMapEntries(CommentedWriter.java:112) ~[?:?]        at space.arim.dazzleconf.ext.snakeyaml.CommentedWriter.writeMap(CommentedWriter.java:93) ~[?:?]
        at space.arim.dazzleconf.ext.snakeyaml.CommentedWriter.lambda$writeValue$0(CommentedWriter.java:188) ~[?:?]
        at space.arim.dazzleconf.ext.snakeyaml.CommentedWriter.descendAndDo(CommentedWriter.java:157) ~[?:?]
        at space.arim.dazzleconf.ext.snakeyaml.CommentedWriter.writeValue(CommentedWriter.java:188) ~[?:?]
        at space.arim.dazzleconf.ext.snakeyaml.CommentedWriter.writeRemainingMapEntries(CommentedWriter.java:112) ~[?:?]        at space.arim.dazzleconf.ext.snakeyaml.CommentedWriter.writeMap(CommentedWriter.java:93) ~[?:?]
        at space.arim.dazzleconf.ext.snakeyaml.CommentedWriter.writeData(CommentedWriter.java:84) ~[?:?]
        at space.arim.dazzleconf.ext.snakeyaml.SnakeYamlConfigurationFactory.writeMapToWriter(SnakeYamlConfigurationFactory.java:148) ~[?:?]
        at space.arim.dazzleconf.factory.AbstractConfigurationFactory$ConfigFactoryDelegate.writeMap(AbstractConfigurationFactory.java:153) ~[?:?]
        at space.arim.dazzleconf.factory.HumanReadableConfigurationFactory.bufferedWriteMap(HumanReadableConfigurationFactory.java:125) ~[?:?]
        at space.arim.dazzleconf.factory.HumanReadableConfigurationFactory.writeMap(HumanReadableConfigurationFactory.java:114) ~[?:?]
        at space.arim.dazzleconf.factory.ConfigurationFormatFactory.write(ConfigurationFormatFactory.java:184) ~[?:?]
        at space.arim.dazzleconf.factory.DelegatingConfigurationFactory.write(DelegatingConfigurationFactory.java:72) ~[?:?]
        at space.arim.dazzleconf.factory.AbstractConfigurationFactory.write(AbstractConfigurationFactory.java:42) ~[?:?]        at space.arim.dazzleconf.helper.ConfigurationHelper.reloadConfigData(ConfigurationHelper.java:102) ~[?:?]
        at pl.minecodes.minespawner.configuration.ConfigurationLoader.reloadConfig(ConfigurationLoader.java:35) ~[?:?]
        at pl.minecodes.minespawner.MineSpawner.loadConfiguration(MineSpawner.java:33) ~[?:?]
        at pl.minecodes.minespawner.MineSpawner.onEnable(MineSpawner.java:41) ~[?:?]
```

**Comment by A248** [comment]
Thanks. This is a bug with the snakeyaml extension. The core should support non-string keys.

Also, I tested that this bug is specific to `CommentMode.alternateWriter` as well as `CommentMode.fullComments`. For now, to avoid it, you can use `CommentMode.headerOnly`, which lacks comment support except for the header.

If you need comments and can't wait until the next release, let me know and I will publish a SNAPSHOT release with the bug fix.

**Comment by A248** [comment]
Released in 1.2.1
